eclair-core/src/main/scala/fr/acinq/eclair/payment/send/PaymentInitiator.scala

```diff
@@ -59,8 +59,12 @@ class PaymentInitiator(nodeParams: NodeParams, router: ActorRef, relayer: ActorR
               sender ! PaymentFailed(paymentId, r.paymentHash, LocalFailure(PaymentSecretMissing) :: Nil)
           }
         case _ =>
-          // NB: we only generate legacy payment onions for now for maximum compatibility.
-          spawnPaymentFsm(paymentCfg) forward SendPayment(r.recipientNodeId, FinalLegacyPayload(r.recipientAmount, finalExpiry), r.maxAttempts, r.assistedRoutes, r.routeParams)
+          val finalPayload = if (r.tlvRecords.isEmpty) {
+            FinalLegacyPayload(r.recipientAmount, finalExpiry)
```

Comment:
I think it's worth updating this line to use the `createSinglePartPayload` function.
This way you can directly provide the additional records without the need for an `if/else`.

Reply:
Okay, this is now using the new `createSinglePartPayload` and the decision between legacy/tlv final payload is made over there, if you don't pass a `payment_secret` it will use the legacy format. As discussed offline this created a bit of confusion for my use case (`keysend` plugin) because it turned out that LND can accept a keysend payment without `payment_secret` (and without the `final_amount` too :warning:) so plugin developers that want to attach custom records to their payment MUST attach a payment secret too.